packages/realm/src/Logger.ts

```diff
@@ -0,0 +1,91 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { assert, binding } from "./internal";
+
+export type LogLevel = "all" | "trace" | "debug" | "detail" | "info" | "warn" | "error" | "fatal" | "off";
+
+export enum NumericLogLevel {
+  All = 0,
+  Trace = 1,
+  Debug = 2,
+  Detail = 3,
+  Info = 4,
+  Warn = 5,
+  Error = 6,
+  Fatal = 7,
+  Off = 8,
+}
+
+export type Logger = (level: NumericLogLevel, message: string) => void;
+
+export type LoggerCallback = (level: LogLevel, message: string) => void;
+
+/** @internal */
+export function toBindingLoggerLevel(arg: LogLevel): binding.LoggerLevel {
+  const bindingLogLevel = inverseTranslationTable[arg];
+  assert(bindingLogLevel !== undefined, `Unexpected log level: ${arg}`);
+  return bindingLogLevel;
+}
+
+/** @internal */
+export function fromBindingLoggerLevelToNumericLogLevel(arg: binding.LoggerLevel): NumericLogLevel {
+  // For now, these map 1-to-1
+  return arg as unknown as NumericLogLevel;
+}
+
+const translationTable: Record<binding.LoggerLevel, LogLevel> = {
+  [binding.LoggerLevel.All]: "all",
+  [binding.LoggerLevel.Trace]: "trace",
+  [binding.LoggerLevel.Debug]: "debug",
+  [binding.LoggerLevel.Detail]: "detail",
+  [binding.LoggerLevel.Info]: "info",
+  [binding.LoggerLevel.Warn]: "warn",
+  [binding.LoggerLevel.Error]: "error",
+  [binding.LoggerLevel.Fatal]: "fatal",
+  [binding.LoggerLevel.Off]: "off",
+};
+
+const inverseTranslationTable: Record<LogLevel, binding.LoggerLevel> = Object.fromEntries(
+  Object.entries(translationTable).map(([key, val]) => [val, Number(key)]),
+) as unknown as Record<LogLevel, binding.LoggerLevel>;
+
+/** @internal */
+export function fromBindingLoggerLevelToLogLevel(arg: binding.LoggerLevel): LogLevel {
+  return translationTable[arg];
+}
+
+const consoleErrorLevels: LogLevel[] = ["error", "fatal"];
+const consoleWarnLevels: LogLevel[] = ["warn"];
```

Comment:
This suggestion is combined with my previous one if we provide the additional log level helpers.
```suggestion
const consoleErrorLevels: ErrorLogLevel[] = ["error", "fatal"];
const consoleWarnLevels: WarnLogLevel[] = ["warn"];
```

In this case, you'd have to change e.g. `consoleErrorLevels.includes(logLevel)` to `consoleErrorLevels.includes(logLevel as ErrorLogLevel)`, so not sure how beneficial this would be for us.

Reply:
Or perhaps `satisfies WarnLogLevel[]`.
```suggestion
const consoleErrorLevels = ["error", "fatal"] satisfies LogLevel[];
const consoleWarnLevels = ["warn"] satisfies LogLevel[];
```

Reply:
@kraenhansen I wrote this as a comment to another related suggestion, but I'm reluctant to go this route (defining `WarnLogLevel` and `ErrorLogLevel`), because it will make the code more difficult to read, and we won't get much out of it.